src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1148,26 +1169,20 @@ public SolutionState AddDocuments(ImmutableArray<DocumentInfo> documentInfos)
                 CheckContainsProject(documentInfosInProject.Key);
                 var oldProject = this.GetProjectState(documentInfosInProject.Key);
 
-                var newDocumentStatesForProjectBuilder = ArrayBuilder<DocumentState>.GetInstance();
+                var newDocumentStatesForProjectBuilder = ArrayBuilder<T>.GetInstance();
 
                 foreach (var documentInfo in documentInfosInProject)
                 {
-                    CheckNotContainsDocument(documentInfo.Id);
-
-                    newDocumentStatesForProjectBuilder.Add(DocumentState.Create(
-                        documentInfo,
-                        oldProject.ParseOptions,
-                        oldProject.LanguageServices,
-                        _solutionServices).UpdateSourceCodeKind(documentInfo.SourceCodeKind));
+                    newDocumentStatesForProjectBuilder.Add(createDocumentState(documentInfo, oldProject));
                 }
 
                 var newDocumentStatesForProject = newDocumentStatesForProjectBuilder.ToImmutableAndFree();
 
-                var newProjectState = oldProject.AddDocuments(newDocumentStatesForProject);
+                var (newProjectState, compilationTranslationAction) = addDocumentsToProjectState(oldProject, newDocumentStatesForProject);
 
                 newSolutionState = newSolutionState.ForkProject(newProjectState,
-                    CompilationTranslationAction.AddDocuments(newDocumentStatesForProject),
-                    newFilePathToDocumentIdsMap: CreateFilePathToDocumentIdsMapWithAddedDocuments(documentInfosInProject.Select(d => newProjectState.GetDocumentState(d.Id))));
+                    compilationTranslationAction,
+                    newFilePathToDocumentIdsMap: CreateFilePathToDocumentIdsMapWithAddedDocuments(newDocumentStatesForProject));
```

Comment:
A couple ways to answer:

1. I would expect it to: the only way that wouldn't happen is if the original state created was then modified by the addition of another document. For example, adding documents in batch is different than adding documents one-by-one. We don't have a case where that would happen and it'd be suspicious to me if that ever went wrong.
2. CreateFilePathToDocumentIdsMapWithAddedDocuments is only consuming the Id and FilePath, so even if a new instance was created, it'd be the same for the purposes of this call. This is trying to avoid creating a bunch of unnecessary allocations. The older approach where we extracted just an id/filepath pair was indeed elegant but seems a bit unnecessary.